### ISSUE TITLE
Common `Service` interface for batch processing.

### ISSUE THREAD

**Issue by hdevalence** [ZcashFoundation/zebra#470]
We want to use futures to allow us to transparently perform batch processing of cryptographic data, as I described in [this post](https://www.zfnd.org/blog/futures-batch-verification/). Batch processing performs less work than the sum total of individual processing, but gives an all-or-nothing answer about the entire batch rather than individual items.  This makes it awkward to use, because conventional APIs require declaring the entire batch upfront, resulting in "transposed" verification logic.  Instead, the new idea is to hide the shared state that allows batch verification behind a `Service` interface that returns a verification future.  When the batch is processed, all the verification futures resolve simultaneously with the verification result.  This means that verification logic can be written generically with regard to batched or individual verification, and without regard to the batch size.

I implemented a first version of this in `ed25519-zebra`, but I'm somewhat unhappy with the interface.  This issue describes the problem and sketches a better alternative, a reusable `tower-batch` middleware.

Batch processing represents a latency-throughput tradeoff. By doing less work overall, throughput increases, but latency also increases, since we need to wait for an entire batch to complete.  When we have a constant stream of requests, this is fine, but when requests come unpredictably, this can result in unbounded latency.  For instance, if requests slow down, requests already waiting for the batch to fill will be "stuck" and complete only much later, or never.  Preventing this requires some policy around maximum batch size and maximum request latency to control when the batch is actually processed.

The existing `ed25519-zebra` implementation addresses the maximum batch size by automatically processing a batch once it gets to a configurable size, and punts on the latency issue by changing its `Request` type to be an enum of `VerificationRequest(...)` and `Flush`, allowing a user of the `Service` to explicitly flush the batch.  This could in principle be hooked up with some other tower layers to have timeouts.  However, there are some problems with this approach.

1.  The flushing policy is split between the service itself (maximum batch size) and the caller (who has to implement some system for latency bounds).  This means that it must be duplicated among all the batch-processing services (signature verification, proof verification, different kinds of proofs, etc.).

2.  Not only does the caller have to implement a system for latency bounds, they also can't do so effectively, since they have to do so on a per-item basis, not on a per-batch basis.  Consider a setting with a batch size of 1000 which is half-filled and then left uncompleted.  A per-item latency bound requires tracking 500 timers instead of one timer, and handling concurrent flush requests if those timers fire at the same time.

3. Because the service takes an enum, it's still not possible to cleanly abstract over batched and individual processing by taking a generic `Service` parameter.

Instead, I think it would be better to have a common `Batch` wrapper similar to the design of `tower::buffer`.  Recall that `tower::buffer` works by moving the wrapped `Service` to a worker task that loops waiting on requests to arrive on an `mpsc` channel before processing them.  The `Buffer` service holds the sender end of the channel and can be cloned (because multi-producer channels can have multiple senders).

We would define
```
enum BatchRequest<R> {
    Request(R),
    Flush,
}

// allows writing req.into() 
impl<R> From<R> for BatchRequest<R> { /* ... */ }

struct Batch<S, R> where S: Service<BatchRequest<R>> { /* ... */ }

impl<S, R> Batch<S, R> where S: Service<BatchRequest<R>> {
    pub fn new(svc: S, max_items: usize, max_latency: std::time::Duration) -> Self {
        /* move svc to worker task */
    }
}

impl<S, R> Service<R> for Batch<S, R> where S: Service<BatchRequest<R>> {
    type Response = S::Response;
    type Error = /* ??? */;
    /* ... */
}
```

To use this, batchable processing is written as a `Service<BatchRequest<R>>`. The `BatchRequest` wrapper provides signaling about when to flush the batch and process its items, moving that responsibility out of each batch processing service and into the common `Batch` wrapper.  Because the `Batch` struct handles batch flushing, it implements `Service<R>`, not `Service<BatchRequest<R>>`, so it can be used to fill a generic parameter without requiring the rest of the code to be batch-aware.

This would be implemented almost identically to `tower::buffer`, with a few differences.  First, rather than having the worker task loop waiting for requests to come in over a channel, it should wait either for requests to come in over a channel or for a timer set when adding the first batch entry. Second, it should keep track of the request count to send a `Flush` request after some batch size is reached.  Third, it should add only minimal buffering to the `mpsc` channel, since batch processing itself is already a kind of buffering.

Our use-case is cryptographic processing, but this problem setting seems more general, so it may be useful to upstream.  For that reason I think it would be good not to write it zebra-specifically if possible, and to add it as a new `tower-batch` crate in our workspace.

**Comment by yaahc** [comment]
I really like this design but I don't understand this point:

> This could in principle be hooked up with some other tower layers to have timeouts. However, there are some problems with this approach.

I don't understand why you can't implement this same system without including the timeout and cloneable service parts into the `BatchRequest` itself. It seems like a service that is a `Buffer<Timeout<FlushPolicy, BatchService<S>>>` should still work with all of the desired features.

Part of this I expect is my lack of familiarity, looking at `Timeout` it looks like it just returns an error if it times out rather than executing a callback, but I feel like this is a short coming of timeout, or maybe it needs to be paired with `Retry` to get it to do the Flush upon a TimeoutError...

Assuming this works the all in one batch service would become:

```rust
type BatchService<S> = Buffer<Retry<FlushPolicy, Timeout<Batch<S>>>>;
```

**Comment by daira** [comment]
> it should wait either for requests to come in over a channel or for a timer set when adding the first batch entry

Is the assumption that all entries have the same maximum latency?

**Comment by hdevalence** [comment]
@yaahc Yes, something like that would probably work, although there might be a few other layers involved to transform the request type into a `BatchRequest`.  However, `Retry` requires a `Clone` service, so the inner service still needs to be `Clone` (each future needs to have a copy of the inner service to be able to send a `Flush` request after timeout).  This means that you'd need to have at least two layers of buffers.  But the big problem is that it requires having one timer per item, rather than one timer per batch.  So at minimum it requires having two layers of message-passing and a timer for each item.

@daira Yes, or more specifically, that there's a maximum latency for each batch's processing.  If some client wants to have tighter latency bounds it can wrap the service in a timeout layer as @yaahc describes.

**Comment by yaahc** [comment]
>  But the big problem is that it requires having one timer per item, rather than one timer per batch. So at minimum it requires having two layers of message-passing and a timer for each item.

Aah, I had thought that putting the `Timeout` behind the `Buffer` would fix that but looking at it I see that the timeout is owned by the future returned by `Timeout` and not part of the service itself.

Also, I think I understand why Retry wont work, it requires the `Request` is clone but we really need a `Retry` layer that produces a new request upon failure so it can send a `Flush`

**Comment by yaahc** [comment]
I'm not sure it's worth it but I'm still thinking about what changes could be made to make this more pluggable. I'm not super familiar with tower and how it all works still but it seems like we could get it to work if we had the following pieces:

* Change `Retry` to have a callback to create the follow up `Request` instead of just cloning the upstream request
* Change `Timeout` to have a callback to get the timeout for the futures it returns from `call` so we could return a handle to a shared timer

With those features I think we would be able to write the BatchService like this

```rust
type BatchService<S> = Retry<FlushOnTimeout, Buffer<Timeout<Batch<S>>>>;
```

But yea I'm beginning to see the scope of the problem. The main reason I'm pushing on this is to keep `BatchService` as simple as possible, I am not enthusiastic about duplicating a bunch of code from other tower services. That said it looks like this is going to require large increases in complexity for everything else and is hard if not impossible to do as many layers, so I think I'm going to stop worrying about this.

**Comment by hdevalence** [comment]
This was closed by #472
